issue_6/contracts/GreenAMM.sol

```diff
@@ -0,0 +1,360 @@
+    // ended up not needing anything else
+}
+
+/**
+@notice 'Engine' of the standard xy=k AMM contract
+ */
+abstract contract GreenAMMEngine is Ownable {
+    bool internal isInitialized;
+
+    /// @dev info about X token
+    TokenData public x;
+    /// @dev info about Y token
+    TokenData public y;
+
+    /// @dev K from X*Y=K
+    /// IMIPORTANT: K has 36 decimals
+    uint256 public k;
+
+    event KChanged(uint256 newK);
+    event Inited(uint256 x, uint256 y);
+    event Minted(uint256 x, uint256 y);
+    event Burned(uint256 x, uint256 y);
+    event SoldX(uint256 x, uint256 y);
+    event SoldY(uint256 x, uint256 y);
+
+    modifier initialized() {
+        require(isInitialized, "not initialized");
+        _;
+    }
+
+    modifier notInitialized() {
+        require(!isInitialized, "already initialized");
+        _;
+    }
+
+    /// @dev asserts that amountX/amountY have same ratio as existing reserves
+    modifier maintainsBalance(uint256 amountX, uint256 amountY) {
+        require(x.reserves != 0 && y.reserves != 0, "Empty pool");
+
+        // since decimals for X and Y are unchanged, don't bother computing *exact* value for x/y:
+        // x1 * dec_x / y1 * dec_y == x2 * dec_x / y2 * dec_y IMPLIES x1/y1 == x2/y2
+        // x1/y1 == x2/y2 IS EQUIVALENT to x1 * y2 == x2 * y1
+        require(x.reserves * amountY == y.reserves * amountX, "Unbalancing");
+        _;
+    }
+
+    /**
+    @notice constructor
+    @param xDecimals decimals of X token
+    @param yDecimals decimals of Y token
+
+    Only 18-decimal tokens are supported
+     */
+    constructor(uint8 xDecimals, uint8 yDecimals) {
+        require(xDecimals == 18, "Only WAD-denominated tokens are supported");
+        require(yDecimals == 18, "Only WAD-denominated tokens are supported");
+        x = TokenData(0);
+        y = TokenData(0);
+    }
+
+    /// @dev recompute K based on new reserve numbers
+    function kRecompute() internal {
+        k = x.reserves * y.reserves; // use increased precision
+        emit KChanged(k);
+    }
+
+    /**
+    @notice 'State Chage' function for 'Init' method
+    @param amountX X tokens to deposit
+    @param amountY Y tokens to deposit
+    @return new K value
+     */
+    function stateChangeInit(uint256 amountX, uint256 amountY) internal notInitialized onlyOwner returns (uint256) {
+        require(amountX != 0 && amountY != 0, "Can't create empty pool");
+
+        isInitialized = true;
+
+        x.reserves += amountX;
+        y.reserves += amountY;
+
+        kRecompute();
+        return k;
+    }
+
+    /**
+    @notice 'State Chage' function for 'Mint' method
+    @param amountX X tokens to deposit
+    @param amountY Y tokens to deposit
+    @return rewardsFraction how many AMM tokens need to be issued (36-decimal fraction of total supply)
+     */
+    function stateChangeMint(uint256 amountX, uint256 amountY)
+        internal
+        initialized
+        maintainsBalance(amountX, amountY)
```

Comment:
It's clearer to use functions instead of modifiers. I would use the latter sparingly.